### ISSUE TITLE
[Bug]: detect credits not working?

### ISSUE THREAD

**Issue by phyzical** [jumoog/intro-skipper#105]
### Describe the bug

cant seem to scan for credits?

i also end up with a 12mb credits.xml file in the cache folder that gets blown away each time i run it?

### Jellyfin installation method

docker

### Container image and tag

jellyfin 10.8.13,  intro skipper latest

### Operating System

docker

### Support Bundle

* Jellyfin version: 10.8.13
* Plugin version: 0.1.16+ccb3bbc6835f
* Queue contents: 91723 episodes, 5565 seasons
* Warnings: `UnableToAddSkipButton, InvalidChromaprintFingerprint`
* FFmpeg: `okay`

FFmpeg version:
```
ffmpeg version 5.1.4-Jellyfin Copyright (c) 2000-2023 the FFmpeg developers
built with gcc 11 (Ubuntu 11.4.0-1ubuntu1~22.04)
configuration: --prefix=/usr/lib/jellyfin-ffmpeg --target-os=linux --extra-libs=-lfftw3f --extra-version=Jellyfin --disable-doc --disable-ffplay --disable-ptx-compression --disable-static --disable-libxcb --disable-sdl2 --disable-xlib --enable-lto --enable-gpl --enable-version3 --enable-shared --enable-gmp --enable-gnutls --enable-chromaprint --enable-libdrm --enable-libass --enable-libfreetype --enable-libfribidi --enable-libfontconfig --enable-libbluray --enable-libmp3lame --enable-libopus --enable-libtheora --enable-libvorbis --enable-libopenmpt --enable-libdav1d --enable-libsvtav1 --enable-libwebp --enable-libvpx --enable-libx264 --enable-libx265 --enable-libzvbi --enable-libzimg --enable-libfdk-aac --arch=amd64 --enable-libshaderc --enable-libplacebo --enable-vulkan --enable-opencl --enable-vaapi --enable-amf --enable-libmfx --enable-ffnvcodec --enable-cuda --enable-cuda-llvm --enable-cuvid --enable-nvdec --enable-nvenc
libavutil      57. 28.100 / 57. 28.100
libavcodec     59. 37.100 / 59. 37.100
libavformat    59. 27.100 / 59. 27.100
libavdevice    59.  7.100 / 59.  7.100
libavfilter     8. 44.100 /  8. 44.100
libswscale      6.  7.100 /  6.  7.100
libswresample   4.  7.100 /  4.  7.100
libpostproc    56.  6.100 / 56.  6.100
```

### Jellyfin logs

```
System.AggregateException: One or more errors occurred. (Input string was not in a correct format.)
 ---> System.FormatException: Input string was not in a correct format.
   at System.Number.ThrowOverflowOrFormatException(ParsingStatus status, TypeCode type)
   at System.Convert.ToInt32(String value, IFormatProvider provider)
   at ConfusedPolarBear.Plugin.IntroSkipper.FFmpegWrapper.DetectBlackFrames(QueuedEpisode episode, TimeRange range, Int32 minimum)
   at ConfusedPolarBear.Plugin.IntroSkipper.BlackFrameAnalyzer.AnalyzeMediaFile(QueuedEpisode episode, AnalysisMode mode, Int32 minimum)
   at ConfusedPolarBear.Plugin.IntroSkipper.BlackFrameAnalyzer.AnalyzeMediaFiles(ReadOnlyCollection`1 analysisQueue, AnalysisMode mode, CancellationToken cancellationToken)
   at ConfusedPolarBear.Plugin.IntroSkipper.BaseItemAnalyzerTask.AnalyzeItems(ReadOnlyCollection`1 items, CancellationToken cancellationToken)
   at ConfusedPolarBear.Plugin.IntroSkipper.BaseItemAnalyzerTask.<>c__DisplayClass5_0.<AnalyzeItems>b__0(KeyValuePair`2 season)
   at System.Threading.Tasks.Parallel.<>c__DisplayClass44_0`2.<PartitionerForEachWorker>b__1(IEnumerator& partitionState, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Parallel.<>c__DisplayClass44_0`2.<PartitionerForEachWorker>b__1(IEnumerator& partitionState, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
   at System.Threading.Tasks.TaskReplicator.Replica.Execute()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.TaskReplicator.Run[TState](ReplicatableUserAction`1 action, ParallelOptions options, Boolean stopOnFirstFailure)
   at System.Threading.Tasks.Parallel.PartitionerForEachWorker[TSource,TLocal](Partitioner`1 source, ParallelOptions parallelOptions, Action`1 simpleBody, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Parallel.ThrowSingleCancellationExceptionOrOtherException(ICollection exceptions, CancellationToken cancelToken, Exception otherException)
   at System.Threading.Tasks.Parallel.PartitionerForEachWorker[TSource,TLocal](Partitioner`1 source, ParallelOptions parallelOptions, Action`1 simpleBody, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.ForEachWorker[TSource,TLocal](IEnumerable`1 source, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.ForEach[TSource](IEnumerable`1 source, ParallelOptions parallelOptions, Action`1 body)
   at ConfusedPolarBear.Plugin.IntroSkipper.BaseItemAnalyzerTask.AnalyzeItems(IProgress`1 progress, CancellationToken cancellationToken)
   at ConfusedPolarBear.Plugin.IntroSkipper.DetectCreditsTask.ExecuteAsync(IProgress`1 progress, CancellationToken cancellationToken)
   at Emby.Server.Implementations.ScheduledTasks.ScheduledTaskWorker.ExecuteInternal(TaskOptions options)
[02:30:54] [INF] [100] Emby.Server.Implementations.ScheduledTasks.TaskManager: Detect Credits Failed after 10 minute(s) and 53 seconds
```

**Comment by jumoog** [comment]
please upload your old xml and the new one here: https://mega.nz/filerequest/2UbcHPkAzjA

**Comment by phyzical** [comment]
@jumoog which xml are you after just the cache credits.xml?

**Comment by rlauu** [comment]
Is there an issue with the xml files? When scanning for intros, the process works smoothly. However, when scanning for credits directly after, the resulting 'credits.xml' file misses items. The only solution I've found is to restart Jellyfin and re-run the task, which generates the complete 'credits.xml' file within seconds. This also occurs in reverse order when scanning credits first, with 'intros.xml' missing items.

**Comment by AbandonedCart** [comment]
Is the file write not being closed?

**Comment by jumoog** [comment]
I'l look into this tomorrow

**Comment by jumoog** [comment]
[ConfusedPolarBear.Plugin.IntroSkipper.zip](https://github.com/jumoog/intro-skipper/files/14857857/ConfusedPolarBear.Plugin.IntroSkipper.zip)

can you test with this ddl? its current 10.9 from git

My old XML files are ~1MB and after scanning they are small but still working. The old xml had alot whitespace.

**Comment by jumoog** [comment]
> @jumoog which xml are you after just the cache credits.xml?

yes please upload the older working 10.8 xml

**Comment by phyzical** [comment]
i haven't actually seen credits skip work personally, i only just switched over when 10.9 unstable was out.

But i had to role back to 10.8 due to subtitles not working in 10.9.

is it possible im using the wrong dll? im using https://github.com/jumoog/intro-skipper/releases/tag/10.9%2F0.1.16.4

Which i now notice has refs to 10.9, ill try 0.1.16.3

**Comment by jumoog** [comment]
You can't use the wrong DLL. 10.8 DLL wont work at all

**Comment by phyzical** [comment]
sorry what i mean is im still running jellyfin 10.8. or should the new dll be backwards compatible ?

**Comment by rlauuzo** [comment]
> [ConfusedPolarBear.Plugin.IntroSkipper.zip](https://github.com/jumoog/intro-skipper/files/14857857/ConfusedPolarBear.Plugin.IntroSkipper.zip)
> 
> can you test with this ddl? its current 10.9 from git
> 
> My old XML files are ~1MB and after scanning they are small but still working. The old xml had alot whitespace.

I tested this DLL, and it caused unexpected behavior. The formatting is missing and all entries are compacted into a single line. Restarting Jellyfin restores the formatted XML with each entry on a separate line. However rescanning the media library removes formatting again.

**Comment by AbandonedCart** [comment]
Once the editor is all cleaned up, it shouldn't be as important. In the meantime, you could probably use one of the minify sites

**Comment by AbandonedCart** [comment]
unminify.com/ does xml

**Comment by phyzical** [comment]
@jumoog have uploaded the failing credits.xml

**Comment by phyzical** [comment]
@AbandonedCart  so i updated to 1.17.0 on jellyfin 10.8.0 this morning, the same issue is now occuring, so i do think its a regession in the code

And after this error i then get "completed scan in 0 seconds" ? and now it will not skip anything, also i cant seem to rollback anymore as there is ionly 1.16 or 1.17 to choose from any idea whats going on?

![image](https://github.com/jumoog/intro-skipper/assets/5182053/53465460-09b9-4453-838c-cb677bdd6bc8)

```
	2024-04-11 12:06:20.475	
[12:06:20] [INF] [16] Emby.Server.Implementations.ScheduledTasks.TaskManager: ExecuteQueuedTasks
	
	
2024-04-11 12:06:20.475	
[12:06:20] [INF] [16] Emby.Server.Implementations.ScheduledTasks.TaskManager: Detect Intros and Credits Failed after 42 minute(s) and 4 seconds
	
	
2024-04-11 12:06:20.475	
   at Emby.Server.Implementations.ScheduledTasks.ScheduledTaskWorker.ExecuteInternal(TaskOptions options)
	
	
2024-04-11 12:06:20.475	
   at ConfusedPolarBear.Plugin.IntroSkipper.DetectIntrosCreditsTask.ExecuteAsync(IProgress`1 progress, CancellationToken cancellationToken)
	
	
2024-04-11 12:06:20.475	
   at ConfusedPolarBear.Plugin.IntroSkipper.BaseItemAnalyzerTask.AnalyzeItems(IProgress`1 progress, CancellationToken cancellationToken)
	
	
2024-04-11 12:06:20.475	
   at System.Threading.Tasks.Parallel.ForEach[TSource](IEnumerable`1 source, ParallelOptions parallelOptions, Action`1 body)
	
	
2024-04-11 12:06:20.475	
   at System.Threading.Tasks.Parallel.ForEachWorker[TSource,TLocal](IEnumerable`1 source, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
	
	
2024-04-11 12:06:20.475	
   at System.Threading.Tasks.Parallel.PartitionerForEachWorker[TSource,TLocal](Partitioner`1 source, ParallelOptions parallelOptions, Action`1 simpleBody, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
	
	
2024-04-11 12:06:20.475	
   at System.Threading.Tasks.Parallel.ThrowSingleCancellationExceptionOrOtherException(ICollection exceptions, CancellationToken cancelToken, Exception otherException)
	
	
2024-04-11 12:06:20.475	
--- End of stack trace from previous location ---
	
	
2024-04-11 12:06:20.475	
   at System.Threading.Tasks.Parallel.PartitionerForEachWorker[TSource,TLocal](Partitioner`1 source, ParallelOptions parallelOptions, Action`1 simpleBody, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
	
	
2024-04-11 12:06:20.475	
   at System.Threading.Tasks.TaskReplicator.Run[TState](ReplicatableUserAction`1 action, ParallelOptions options, Boolean stopOnFirstFailure)
	
	
2024-04-11 12:06:20.475	
   --- End of inner exception stack trace ---
	
	
2024-04-11 12:06:20.475	
   at System.Threading.Tasks.TaskReplicator.Replica.Execute()
	
	
2024-04-11 12:06:20.475	
   at System.Threading.Tasks.Parallel.<>c__DisplayClass44_0`2.<PartitionerForEachWorker>b__1(IEnumerator& partitionState, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
	
	
2024-04-11 12:06:20.475	
--- End of stack trace from previous location ---
	
	
2024-04-11 12:06:20.475	
   at System.Threading.Tasks.Parallel.<>c__DisplayClass44_0`2.<PartitionerForEachWorker>b__1(IEnumerator& partitionState, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
	
	
2024-04-11 12:06:20.475	
   at ConfusedPolarBear.Plugin.IntroSkipper.BaseItemAnalyzerTask.<>c__DisplayClass5_0.<AnalyzeItems>b__0(KeyValuePair`2 season)
	
	
2024-04-11 12:06:20.475	
   at ConfusedPolarBear.Plugin.IntroSkipper.BaseItemAnalyzerTask.AnalyzeItems(ReadOnlyCollection`1 items, CancellationToken cancellationToken)
	
	
2024-04-11 12:06:20.475	
   at ConfusedPolarBear.Plugin.IntroSkipper.BlackFrameAnalyzer.AnalyzeMediaFiles(ReadOnlyCollection`1 analysisQueue, AnalysisMode mode, CancellationToken cancellationToken)
	
	
2024-04-11 12:06:20.475	
   at ConfusedPolarBear.Plugin.IntroSkipper.BlackFrameAnalyzer.AnalyzeMediaFile(QueuedEpisode episode, AnalysisMode mode, Int32 minimum)
	
	
2024-04-11 12:06:20.475	
   at ConfusedPolarBear.Plugin.IntroSkipper.FFmpegWrapper.DetectBlackFrames(QueuedEpisode episode, TimeRange range, Int32 minimum)
	
	
2024-04-11 12:06:20.475	
   at System.Convert.ToInt32(String value, IFormatProvider provider)
	
	
2024-04-11 12:06:20.475	
   at System.Number.ThrowOverflowOrFormatException(ParsingStatus status, TypeCode type)
	
	
2024-04-11 12:06:20.475	
 ---> System.FormatException: Input string was not in a correct format.
	
	
2024-04-11 12:06:20.475	
System.AggregateException: One or more errors occurred. (Input string was not in a correct format.)
	
	
2024-04-11 12:06:20.475	
[12:06:20] [ERR] [16] Emby.Server.Implementations.ScheduledTasks.TaskManager: Error
```

**Comment by AbandonedCart** [comment]
A regression in the code would mean the cause of an issue was identified, fixed, and then some change caused it to break again. We haven't been able to identify what is causing your issue or why you are the only one seeming to have it.

**Comment by phyzical** [comment]
i think the current problem i'm facing about atleast re running only intro scanner relates to (i think) if the AnalyzeItems fails then  `Plugin.Instance!.AnalyzerTaskIsRunning `stays as false so every run after it thinks its still running when in reailty it just failed 

```
 baseCreditAnalyzer.AnalyzeItems(progress, cancellationToken);
 Plugin.Instance!.AnalyzerTaskIsRunning = false;
```

**Comment by AbandonedCart** [comment]
So what is the log attached to the previous post? That says it ran for about 42 minutes and the issue is with a string.

**Comment by phyzical** [comment]
Ah sorry heres the other parts 
![image](https://github.com/jumoog/intro-skipper/assets/5182053/489a9623-8146-4ae9-94d3-be1581ea222e)

**Comment by phyzical** [comment]
Okay So i can confirm, i think the 0 second bug is related to not resetting the AnalyzerTaskIsRunning when AnalyzeItems fails, ill look at making a pr later tonight to stop that.

after running the intro scan on its own, skipping intros is back so i guess the combined script just blatted my stored intros or something.

Is there anything you would like me to try to narrow down whats going on my end with credit scans?

**Comment by AbandonedCart** [comment]
We could try narrowing down an actual **cause**. A try / catch is meant to prevent unexpected errors from taking down an entire process, not as a band-aid for an error we know is there and don't want to actually fix.

**Comment by phyzical** [comment]
>Is there anything you would like me to try to narrow down whats going on my end with credit scans?

**Comment by AbandonedCart** [comment]
The reason AnalyzeItems fails

**Comment by phyzical** [comment]
im asking how i can do that.

i tried adding a log statement

```
            Logger?.LogError("BLACKFRAME {StrPercent} {StrTime}", strPercent, strTime);
            
           new BlackFrame(
                             Convert.ToInt32(strPercent, CultureInfo.InvariantCulture),
                             Convert.ToDouble(strTime, CultureInfo.InvariantCulture));
```

and the trace looked like the following, i see nothing wrong with the last logged message around BLACKFRAME, so not sure why the int conversion fails

![image](https://github.com/jumoog/intro-skipper/assets/5182053/801d0a41-3e20-4dc6-a143-7b8a47bc753e)

**Comment by phyzical** [comment]
Okay so if i try and catch the Blackframe and skip the affected file/s everything clears fine. 

I am then also able to autoskip / skip credits.

The only other issue i seem to face is if it skips to the end of the media then it doesnt autoplay the next file.

Should we instead make sure the max skip is say 3~ seconds before the end of the playback maybe make this configurable?

**Comment by AbandonedCart** [comment]
I guess I'm hesitant because your setup seems to have bugs and trying to shape the plugin around those bugs carries just as much risk of introducing bugs to others that didn't have them.

I think the first priority should be figuring out why your particular instance of the plugin is different from the normal setup and work back from there to determine if the plugin needs changing or it's something in your setup.

The error you are getting says that the value you have saved for percentage can't be parsed to an int correctly. The first thing to look at would seem to be if changing the value changes anything.

**Comment by phyzical** [comment]
So i added context to spit out the affecting files. it turned out to be a single file, after removing this file the scan works as expected.

This file though has no playback issues as far as i can see.

i still think the whole scan shouldn't be killed when a single file fails though, i cant seem to see why it doesnt like the value passed to convert to int...

**Comment by AbandonedCart** [comment]
What's the file? I agree that shouldn't happen, but a blanket catch seems like it could allow errors that need to be errors to get overlooked

**Comment by phyzical** [comment]
https://mega.nz/file/Iq9V2DRR#7o9kxdZrf_8NCIIONcJE5fy_tP4oDuP30yHzp6evfWc

**Comment by jumoog** [comment]
Okay i can reproduce this

**Comment by jumoog** [comment]
fixed in d288a2b5cf3399717652add648b7831d6edcc280

**Comment by phyzical** [comment]
👍 ha no wonder it was such a rare case thanks for fixing!
